cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/partitioned/PartitionedFileSetDataset.java

```diff
@@ -171,53 +170,72 @@ public void postTxCommit() {
   public boolean rollbackTx() throws Exception {
     // rollback all the partition add and drop operations, in reverse order
     // if any throw exception, suppress it temporarily while attempting to roll back the remainder operations
-    IOException ioe = null;
+    Exception caughtExn = null;
     for (int i = operationsInThisTx.size() - 1; i >= 0; i--) {
-      PathOperation operation = operationsInThisTx.get(i);
+      PartitionOperation operation = operationsInThisTx.get(i);
       try {
         rollbackOperation(operation);
-      } catch (IOException e) {
-        if (ioe == null) {
-          ioe = e;
+      } catch (Exception e) {
+        if (caughtExn == null) {
+          caughtExn = e;
         } else {
-          ioe.addSuppressed(e);
+          caughtExn.addSuppressed(e);
         }
       }
     }
-    if (ioe != null) {
-      throw ioe;
+    if (caughtExn != null) {
+      throw caughtExn;
     }
 
     this.tx = null;
     return super.rollbackTx();
   }
 
-  private void rollbackOperation(PathOperation operation) throws IOException {
+  private void rollbackOperation(PartitionOperation operation) throws Exception {
     switch (operation.getOperationType()) {
       case CREATE:
-        undoPartitionCreate(operation.getRelativePath());
+        undoPartitionCreate(operation.getPartitionKey(), operation.getRelativePath());
         break;
       case DROP:
-        undoPartitionDelete(operation.getRelativePath());
+        undoPartitionDelete(operation.getPartitionKey(), operation.getRelativePath());
         break;
       default:
         // this shouldn't happen, since we only have the above types defined in the Enum.
         throw new IllegalArgumentException("Unknown operation type: " + operation.getOperationType());
     }
   }
 
-  private void undoPartitionDelete(String relativePath) throws IOException {
+  private void undoPartitionDelete(PartitionKey partitionKey, String relativePath) throws Exception {
     // move from quarantine, back to original location
     Location srcLocation = getQuarantineLocation().append(relativePath);
     if (srcLocation.exists()) {
       srcLocation.renameTo(files.getLocation(relativePath));
     }
+    // recreating the partition in Hive only makes sense if the rename succeeds
+    addPartitionToExplore(partitionKey, relativePath);
   }
 
-  private void undoPartitionCreate(String relativePath) throws IOException {
-    Location location = files.getLocation(relativePath);
-    if (location.exists() && !location.delete(true)) {
-      throw new IOException(String.format("Failed to delete location %s.", location));
+  private void undoPartitionCreate(PartitionKey partitionKey, String relativePath) throws Exception {
+    Exception caughtExn = null;
+    try {
+      dropPartitionFromExplore(partitionKey);
```

Comment:
Ah ok. I brought this up, because that method ([addPartition](https://github.com/caskdata/cdap/pull/7162/files/b5199fc750f822c9fac46bb66a76bfb3d78d1d67#diff-3c5d2259d081358844e619de52b3741aR273)) accepts a parameter whether to enable explore on a per-partition basis. However, it seems like we always only pass in true for the explore parameter.